rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,417 @@
+
+#### Upload Queue
+
+The Upload Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending upload jobs.
+The upload jobs are used by the upload service to know where to download the file from before uploading them to S3.
+For the first implementation, the job sleep time will be 15min. Eventually, the job sleep time will depend on the size
+of the file being uploaded.
+
+Cloudwatch events can be used to monitor the SQS length. Additional computing resource can be spawned in response to an
+increase in jobs in the queue.
+
+##### Upload Job Entry
+
+These are the fields that will be in the upload job placed in the queue:
+
+- Link - the link to file to download. Use to retrieve the file from DropBox.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location.
+- file_name - the name of the file being downloaded.
+
+#### Upload Service
+
+A computing resource within the AWS cloud. It will receive upload jobs from the Upload Queue. Those jobs contain
+the information needed by the computing resource to download a file from DropBox and upload it to S3. The computing resource
+will also verify the integrity of the download and upload. The computing resource will also change the state of the
+upload in the Upload table as needed.
+
+While the upload job is processing, the computing resource with occasionally poll the upload table for a cancel
+pending status. If the cancel pending status is set, the computing resource will delete any data uploaded to S3,
+remove the upload job from the queue, and change the status of the upload to canceled.
+
+If an unrecoverable error occurs while processing the upload, the computing resource will return the upload job to the
+queue and update the upload table.
+
+If an error is encountered that cannot be fixed by retrying, the upload status will change to failed. The error can be
+returned
+
+##### Upload Table
+
+The upload table is a DynamoDB table that tracks the current status of uploads.
+
+DynamoDB was favored over adding additional fields to the existing Database. The upload table will be updated and queried
+frequently by the browser app and the upload service. The browser app will query periodically for upload progress.
+The upload service will query the upload table for a cancel pending status, and to update the progress and status of
+the upload. With the current database cached behind CloudFront, the upload service would constantly need to clear the
+CloudFront cache if the existing database was used.
+
+The table will have the following columns:
+
+- dataset_id - primary key. This field is used to identify the upload job.
+- status - the current state of the upload. This field is updated to inform the user of the current state of the upload.
```

Comment:
This is where you would indicate that status is one of the values in the state table.

Reply:
added a reference to the status table